### ISSUE TITLE
primitives should not trigger the no-param-reassign error

### ISSUE THREAD

**Issue by skyleguy** [eslint/eslint#13935]
**What rule do you want to change?**
no-param-reassign
**Does this change cause the rule to produce more or fewer warnings?**
fewer
**How will the change be implemented? (New option, new default behavior, etc.)?**
type checking of param (string | number)
**Please provide some example code that this change will affect:**

```js
public replaceDoubleQuoteWithSingleQuote(input: string): string {
  input = input.replace('"', "'");
  return input;
}
```

**What does the rule currently do for this code?**
Throws error for no-param-reassign even though primitive values like strings and numbers do not have any danger of reassigning the variable from where the method was called.
**What will the rule do after it's changed?**
There will be no error with the previous method
**Are you willing to submit a pull request to implement this change?**
Not allowed to, unfortunately

**Comment by mdjermanovic** [comment]
Hi @skyleguy, thanks for the issue!

ESLint core rules don't have type info, so it wouldn't be possible to implement this logic in the core `no-param-reassign` rule. Type-aware rules can be implemented in [@typescript-eslint/eslint-plugin](https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/eslint-plugin). 

> Throws error for no-param-reassign even though primitive values like strings and numbers do not have any danger of reassigning the variable from where the method was called.

This rule disallows reassigning local param variables, what's the difference between primitive and non-primitive values in that regard?

**Comment by skyleguy** [comment]
Hey, I guess I can try to open this in the plugin's github, sorry! 

As far as I can see this rule really exists to prevent developers from accidentally modifying variables they dont intend to. For example if i have an object with a field called 'name' and pass that object into a method where the name field is changed, it changes the value even from where the method was called because it passed the object by reference.

```
export class Thing: {
   name: string;
  constructor(name: string) {
    this.name = name;
  }
}

public main() {
  const thing: Thing = new Thing('Joe');
  this.changeThingName(thing); // even though i didnt assign thing to the return value from changeThingName() the value was still modified.
  console.log(thing.name) // logs 'Billy' to the console
}

public changeThingName(thing: Thing): Thing {
  thing.name = 'Billy';
  return thing;
}
```

But numbers and strings are passed by value so this issue does not have to be worried about with them

**Comment by mdjermanovic** [comment]
> ```ts
> public changeThingName(thing: Thing): Thing {
>   thing.name = 'Billy';
>   return thing;
> }
> ```

Does allowing `thing = something` when `thing` is a primitive but disallowing `thing = something` when `thing` is an object helps to avoid this? It looks to me more like a request to allow assignments to variables but not assignments to properties.

**Comment by skyleguy** [comment]
I do think it would help to avoid it but yeah that could definitely be a better way to word it. 

because doing:

```js
public main() {
  const thing: string = 'Joe';
  this.changeName(thing);
  console.log(thing) // logs 'Joe' to the console
}

public changeName(input: string): string {
  input = input.toUpperCase();
  return thing;
}
```
should not throw an eslint error like the previous one because there is not any danger to modifying something by accident with primitives like there is with objects. It could also be that there is another reason this rule exists that I am being completely oblivious too but it seems the main reason for this error is to avoid messing up objects by accident

**Comment by mdjermanovic** [comment]
This rule disallows two patterns:

* Reassignments of param variables, e.g., `param = foo;`
* Modifications of params, e.g., `param.foo = bar;`. This is optional, `"props": true` should be set.

If I correctly understand the type checking proposal, the behavior would be:

```js
(param1: string, param2: Thing, param3: string, param4: Thing) => {
    param1 = foo; // should be ok, because it's a primitive
    param2 = bar; // still not ok, because it's an object
    param3.prop = baz; // not ok
    param4.prop = quux; // not ok
}
```

Reassignments don't modify objects, so it looks to me there's no difference between `param: string` and `param: Thing` in that regard, but I'm maybe missing something.

> Hey, I guess I can try to open this in the plugin's github, sorry!

No problem! The repo for @typescript-eslint/eslint-plugin issues is:

https://github.com/typescript-eslint/typescript-eslint
